File: backend/ProjectWorker-Execution-Service/src/main/java/com/example/ProjectWorker_Execution_Service/service/ContractService.java
package com.example.ProjectWorker_Execution_Service.service;

import com.example.ProjectWorker_Execution_Service.dto.ContractPartyDetailsRequest;
import com.example.ProjectWorker_Execution_Service.dto.ContractResponse;
import com.example.ProjectWorker_Execution_Service.security.UserPrincipal;

import java.util.List;

public interface ContractService {

    ContractResponse getOrCreateForProject(String projectId, ContractPartyDetailsRequest details, UserPrincipal principal);

    List<ContractResponse> getMyContracts(UserPrincipal principal);

    ContractResponse sign(String contractId, UserPrincipal principal);

    List<ContractResponse> getAllContracts();

    ContractResponse validate(String contractId);
}

File: frontend/lib/contractService.ts
import executionApi from "./executionApi";

export interface ContractResponse {
  id: string;
  projectId: string;
  projectTitle: string;

  workerId: string;
  workerName: string;
  workerEmail: string | null;
  workerPhone: string | null;

  providerId: string;
  providerName: string;
  providerEmail: string | null;
  providerPhone: string | null;

  workerSigned: boolean;
  workerSignedAt: string | null;
  providerSigned: boolean;
  providerSignedAt: string | null;
  adminValidated: boolean;
  validatedAt: string | null;
  createdAt: string;
}

export const contractService = {
  async getMyContracts(): Promise<ContractResponse[]> {
    const res = await executionApi.get<ContractResponse[]>("/api/contracts/my");
    return res.data;
  },

  async getOrCreateForProject(projectId: string, details?: {
    workerName?: string; workerEmail?: string; workerPhone?: string;
    providerName?: string; providerEmail?: string; providerPhone?: string;
  }): Promise<ContractResponse> {
    const res = await executionApi.post<ContractResponse>(`/api/contracts/project/${projectId}`, details ?? {});
    return res.data;
  },

  async sign(contractId: string): Promise<ContractResponse> {
    const res = await executionApi.patch<ContractResponse>(`/api/contracts/${contractId}/sign`);
    return res.data;
  },

  async getAllContracts(): Promise<ContractResponse[]> {
    const res = await executionApi.get<ContractResponse[]>("/api/contracts/all");
    return res.data;
  },

  async validate(contractId: string): Promise<ContractResponse> {
    const res = await executionApi.patch<ContractResponse>(`/api/contracts/${contractId}/validate`);
    return res.data;
  },
};
